keps/83-workload-preemption/README.md

```diff
@@ -0,0 +1,666 @@
+field to true.
+2. The Kubernetes' job controller deletes the Job's Pods.
+3. The Kubelets send SIGTERM signals to the Pod's containers, which can have a
+   graceful termination logic.
+
+This implies the following:
+- Pods of a workload could implement checkpointing as part of their graceful
+  termination.
+- The resources from these Pods are not immediately available and they could
+  be arbitrarily delayed.
+- While Pods are terminating, a ClusterQueue's quota could be oversubscribed.
+
+The kubernetes Job status includes the number of Pending/Running Pods that are
+not terminating (don't have a `.metadata.deletionTimestamp`). We could use
+this information and write the old admission spec into an annotation to keep
+track of usage from non-terminating Pods. But this will be left for future work.
+
+### Increased admission latency
+
+Calculating and executing preemption is expensive. Potentially, every
+workload might benefit from preemption of running Workloads.
+
+To mitigate this, we will keep track of the minimum priority among the running
+Workloads in a ClusterQueue and cohort. If the minimum priorities are higher
+than or equal to the incoming Workload, we will skip preemption for it
+altogether.
+
+The assumption is that workloads with low priority are more common than
+workloads with higher priority and that Workloads are sent to ClusterQueues
+where most Workloads have the same priority.
+
+Additionally, the preemption algorithm is mostly a linear pass over the running
+workloads (plus sorting), so it doesn't add a significant complexity overhead
+over building the scheduling snapshot every cycle.
+
+The API updates from preemption will be executed in parallel.
+
+## Design Details
+
+The proposal consists of new API fields and a preemption algorithm.
+
+### ClusterQueue API changes
+
+The new API fields in ClusterQueue describe how to influence the selection
+of Workloads to preempt.
+
+```golang
+type ClusterQueueSpec struct {
+  ...
+  // preemption describes policies to preempt Workloads from ClusterQueues.
+  preemption ClusterQueuePreemption
+}
+
+type PreemptionByLowerPriorityWhenBorrowingPolicy string
+
+const (
+  PreemptionByLowerPriorityWhenBorrowingAllow = "Allow"
+  PreemptionByLowerPriorityWhenBorrowingDisallow = "Disallow"
+)
+
+type ClusterQueuePreemption struct {
+  // waitBeforePreemptionSeconds is the time in seconds that the ClusterQueue
+  // is willing to wait for Workloads to finish since the first time the
+  // Workload was attempted for admission, before preempting active Workloads
+  // with lower priority.
+  // The time when the Workload was first attempted for admission is present
+  // in the transitionTimestamp of the Admitted condition.
+  // If null, it means to wait indefinitely, effectively disabling preemption
+  // for incoming Workloads in this ClusterQueue.
+  WaitBeforePreemptionSeconds *int64
```

Comment:
Jordan and I spend some time thinking about this field, beyond just the name.

Here is one scenario where things get confusing:

Setup:
1. low priority stuff is running and consuming all resources, mode=StrictFIFO, WaitBeforePreemptionSeconds=2 seconds
2. add medium priority workload b at t=0
3. add high priority workload a at t=1
4. one second passes (t=2)
5. queue looks like high(workload-a(waiting 1 second)), medium(workload-b(waiting 2 seconds))

What _should_ happen?

In best-effort mode:
* we would look at workload-a, skip preemption (because it hadn't been waiting long enough)
* continue to workload-b, trigger preemption of things matching b's resource requirements
* change in available resources (because of our preemption) would make us reevaluate the whole list
* if workload-a fit within the space made for workload-b by preemption, we would run it, otherwise, we'd run workload-b

In strict mode (currently):
* we would only look at workload-a, and not preempt anything because it had not been waiting long enough
* this means even though workload-b *has* been waiting longer than WaitBeforePreemptionSeconds, nothing triggers the queue to preempt to make progress until workload-a reaches that threshold

Naively, a user would expect once something in the queue had reached the wait period to consider preemption, we would start making room for the queue to progress to running that workload, even if that transitively meant making room for the workload at HEAD in strict mode.

I think we can make that change relatively easily by not just looking at the wait time for the head of the queue, but the longest wait time we have in the queue.

As for the field name, `triggerAfterWorkloadWaitingSeconds` should do.

Reply:
I am having second thoughts on this parameter. Without informing the victim job of the upcoming preemption, how is shifting the time we preempt going to be useful at the limit?

Reply:
Going with the interpretation of "at the limit" as when all the ClusterQueues are full, thus we would require preemption in all scenarios.

The time is measured from the first time we attempt to admit a Workload.
In StrictFIFO (assuming no higher priority workloads jump to the head of the queue), this means that we still wait before preemption after each admission.
BestEffortFIFO is more complicated, because the head will keep changing as long as there are no other changes in the cluster (like workloads finishing). In an extreme case, we would go through the entire queue and start the timers for all the workloads. Once the timer is fulfilled, we will do preemptions for all the elements in the queue until they are all admitted. Arguably, this is less useful.

I guess we could park it and instead provide a knob to control whether to preempt within a ClusterQueue or not, along side with the knob to preempt within the cohort.

Reply:
I think we can park it for now. It feels like in most cases this will simply shift the start of the incoming job, which is equivalent to as if the incoming job was submitted `WaitBeforePreemptionSeconds` later in most cases.

To do "cooperative" preemption, it is more useful to find ways to improve victim selection, probably via an explicit signal from the workloads (similar to preemption cost). For example, in the case of v1.Job and completions > parallelism, we can calculate how much progress it did by looking at success count and completions and prefer less the ones that are almost done?

Reply:
Moved the field into the "Alternatives" section for now.